File: 15_Greedy/05_Hand_of_Straights/0846-hand-of-straights.cpp
/*
Problem: LeetCode 846 - Hand of Straights

Description:
Alice has a hand of cards, given as an array of integers `hand`, where `hand[i]` is the value of the ith card.
A valid hand is a hand where every group of W cards is made up of cards of the same value.
Return true if and only if she can reorder the cards in her hand to form a valid hand.

Intuition:
To check if Alice can form valid groups of cards, we can use a greedy approach.
We can sort the cards in ascending order and then try to form groups of size W.

Approach:
1. Create a map to store the frequency of each card in the hand.
2. Sort the hand in ascending order.
3. Iterate through the sorted hand and for each card, try to form a group of size W.
4. If the current card frequency in the map is greater than 0, decrement its frequency by 1.
   Then, check if there are W-1 consecutive cards with frequencies greater than 0 after this card.
   If true, decrement the frequencies of these consecutive cards by 1 to form a group.
5. If we can't form a group of size W, return false.
6. If all groups are successfully formed, return true.

Time Complexity:
The time complexity is O(n log n) due to sorting the hand, where n is the number of cards in the hand.

Space Complexity:
The space complexity is O(n) to store the card frequencies in the map.
*/

class Solution {
  public:
    bool isNStraightHand(vector<int> &hand, int W) {
        if (hand.size() % W != 0) {
            return false; // If the hand size is not divisible by W, can't form valid groups
        }

        map<int, int> cardFreq; // Map to store the frequency of each card

        for (int card : hand) {
            cardFreq[card]++;
        }

        sort(hand.begin(), hand.end()); // Sort the hand in ascending order

        for (int card : hand) {
            if (cardFreq[card] > 0) {
                for (int i = 0; i < W; i++) {
                    if (cardFreq[card + i] == 0) {
                        return false; // Can't form a group of size W
                    }

                    cardFreq[card + i]--;
                }
            }
        }

        return true; // All groups of size W are formed successfully
    }
};

/*
// Beats 99% Runtime and Memory
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        int n = hand.size();
        // Check if the hand size is divisible by the groupSize
        if (n % groupSize != 0)
            return false;

        // Sort the hand array in ascending order
        sort(hand.begin(), hand.end());

        // Iterate through the hand array
        for (int i = 0; i < n; i++) {
            // Skip elements that have been marked as used (-1)
            if (hand[i] == -1)
                continue;

            int k = i;
            // Try to find the next groupSize consecutive elements
            for (int j = 1; j < groupSize; j++) {
                // Search for the next consecutive element by incrementing k
                while (k < n && hand[i] + j != hand[k])
                    k++;

                // If k reaches the end or the next element is not found, return false
                if (k == n)
                    return false;

                // Mark the found element as used by setting it to -1
                hand[k] = -1;
            }
        }

        // If all groups are found successfully, return true
        return true;
    }
};
*/

File: 15_Greedy/05_Hand_of_Straights/0846-hand-of-straights.rs
/*
Problem: LeetCode 846 - Hand of Straights

Key Idea:
The key idea is to group the cards into consecutive hands of 'W' cards each, where 'W' is the given group size, such that each group is consecutive and all groups are of the same size.

Approach:
1. Create a HashMap to count the frequency of each card in the input hand.
2. Sort the unique cards from the input hand.
3. Iterate through the sorted unique cards:
   - For each card, check if its frequency is greater than zero.
   - If so, try to form a group of 'W' consecutive cards starting from the current card.
   - To do this, iterate from the current card to the next 'W - 1' cards, reducing their frequencies.
   - If a group of 'W' consecutive cards cannot be formed, return false.
4. If all groups can be formed, return true.

Time Complexity:
O(N * log(N)), where N is the number of unique cards in the input hand. This is because we need to sort the unique cards.

Space Complexity:
O(N), where N is the number of unique cards in the input hand. This is for the HashMap to store the frequencies of the cards.
*/

use std::collections::HashMap;

impl Solution {
    pub fn is_n_straight_hand(hand: Vec<i32>, w: i32) -> bool {
        if hand.len() % w as usize != 0 {
            return false;
        }

        let mut freq_map: HashMap<i32, i32> = HashMap::new();

        // Count the frequency of each card in the hand
        for card in &hand {
            *freq_map.entry(*card).or_insert(0) += 1;
        }

        let mut unique_cards: Vec<i32> = freq_map.keys().cloned().collect();
        unique_cards.sort();

        for card in unique_cards {
            let freq = *freq_map.get(&card).unwrap();

            if freq > 0 {
                for i in 0..w {
                    let next_card = card + i;

                    if let Some(&next_freq) = freq_map.get(&next_card) {
                        if next_freq < freq {
                            return false; // Cannot form a group
                        }
                        *freq_map.get_mut(&next_card).unwrap() -= freq;
                    } else {
                        return false; // Cannot form a group
                    }
                }
            }
        }

        true
    }
}

/*
// Using heap

use std::collections::{BinaryHeap, HashMap};
use std::cmp::Reverse;

impl Solution {
    pub fn is_n_straight_hand(hand: Vec<i32>, group_size: i32) -> bool {
        if hand.len() % group_size as usize != 0 {
            return false;
        }

        let mut card_counts: HashMap<i32, i32> = HashMap::new();
        for card in hand {
            *card_counts.entry(card).or_insert(0) += 1;
        }

        let mut min_heap: BinaryHeap<Reverse<i32>> = BinaryHeap::new();
        for &card in card_counts.keys() {
            min_heap.push(Reverse(card));
        }

        while let Some(Reverse(first_card)) = min_heap.peek() {
            for card in *first_card..(*first_card + group_size) {
                if !card_counts.contains_key(&card) {
                    return false;
                }

                if let Some(&count) = card_counts.get(&card) {
                    if count == 1 {
                        if let Some(Reverse(peeked_card)) = min_heap.peek() {
                            if card != *peeked_card {
                                return false;
                            }
                            min_heap.pop();
                        }
                    } else {
                        *card_counts.get_mut(&card).unwrap() -= 1;
                    }
                }
            }
        }

        true
    }
}
*/
